Marlin/src/gcode/motion/G2_G3.cpp

```diff
@@ -179,7 +185,7 @@ void plan_arc(
     // Update raw location
     raw[p_axis] = center_P + r_P;
     raw[q_axis] = center_Q + r_Q;
-    raw[l_axis] += linear_per_segment;
+    raw[l_axis] = start_L;
```

Comment:
I don't understand what you are trying to communicate.   The problem UBL is seeing is the Z-Correction is being applied again and again and is cumulative.   By resetting raw[l_axis] to the starting value of Z the code behave properly.   Do we need some extra logic to handle your case?

Reply:
This change breaks arcs that have movement in the linear axis.

Reply:
OK.    But for sure the code is broken and doesn't work right in the normal Z-Axis.
Perhaps it needs more work.   But the original Arc function got broken by the Linear Axis changes.
I don't know what has to be done to fix the Linear Axis side of this.

Reply:
I think the root cause of the problem is the Z-Height Correction is being applied at a different time between the different leveling systems.    I copied what the original Mesh Leveling did.   It would take the current Z, add on the correction, and then call the planner::.   

But in the case where l_axis == Z_AXIS, we can't keep incrementing the Z Axis for each segment.   That causes the 'Air Print' mentioned here: https://github.com/MarlinFirmware/Marlin/pull/12386#issuecomment-437705604

Right now...  I have the changes wrapped with #if ENABLED(UBL).
```
    #if ENABLED(AUTO_BED_LEVELING_UBL)
      raw[l_axis] = start_L;
    #else
      raw[l_axis] += linear_per_segment;
    #endif
```
For the time being...   Could we add one more qualifier to resolve the issue???   something like:
```
    #if ENABLED(AUTO_BED_LEVELING_UBL) && ENABLED(CARTESIAN)
      raw[l_axis] = start_L;
    #else
      raw[l_axis] += linear_per_segment;
    #endif
```

Reply:
That would be okay, but we should figure out a proper resolution soon. The way other leveling systems work, the Z adjustment is transparent to the high-level code, but the Z that gets passed on to the Z stepper in terms of steps gets adjusted. (Same goes for XY with the planar matrix-based systems.) If UBL could move its Z adjustments to `planner.apply_leveling`/`unapply_leveling` for all cases —including least-segmented Cartesian— but still preserve its optimizations for straight lines and so on, that would give us the best of both worlds.